guide.md

```diff
@@ -0,0 +1,559 @@
+to write builds that are _incorrect_ and fail to declare all their dependencies.
+
+Nix itself is fairly old (the first commit dates back to 2003) and as old
+software tends to do, there are "legacy" and "modern" ways of doing certain
+things that are both still in common use. When defining development shells and
+setting up projects for use with Nix, we choose the newer "flake" method over
+the legacy method (involving a `shell.nix` file) that you may run into in other
+places.
+
+A _flake_, defined in a file conventionally called `flake.nix`, can be thought
+of as a self-contained "Nix module". For our purposes, it will suffice to think
+of it as a way to define a cross-platform development shell, with the versions
+of the _inputs_ to the shell pinned in a lockfile (`flake.lock`). To get
+started, put the following code in a file named `flake.nix` at the root of the
+repository:
+
+```nix
+{
+  inputs = {
+    # Tracks the nixos-22.05 tag on the nixpkgs repo.
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.05";
+
+    # The flake format itself is very minimal, so the use of this
+    # library is common.
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  # Here we can define various kinds of "outputs": packages, tests, 
+  # and so on, but we will only define a development shell.
+
+  outputs = { nixpkgs, flake-utils, ... }:
+
+    # For every platform that Nix supports, we ...
+    flake-utils.lib.eachDefaultSystem (system:
+
+      # ... get the package set for this particular platform ...
+      let pkgs = import nixpkgs { inherit system; };
+      in 
+      {
+
+        # ... and define a development shell for it ...
+        devShells.default =
+
+          # ... with no globally-available CC toolchain ...
+          pkgs.mkShellNoCC {
+            name = "rules_nixpkgs_shell";
+
+            # ... which makes available the following dependencies, 
+            # all sourced from the `pkgs` package set:
+            packages = with pkgs; [ bazel_5 bazel-buildtools cacert nix git ];
+          };
+      });
+}
+```
+
+It's not necessary at this juncture to understand the details of the Nix code
+above. (For those curious, [this][flake-post] series of articles on the Tweag
+blog is a good start.) It's worth noting the `packages = ...` line, however. The
+square brackets, as you might expect, set off a _list_ in Nix, with elements
+separated (admittedly a bit oddly) by whitespace. If you need packages in your
+development shell that aren't listed there already (compilers, development
+tools, editors, anything at all) you can add them to the `packages` list, and
+they'll become available within the shell when you enter it. In order to search
+for the package you want in Nixpkgs, there is a [package
+search](https://search.nixos.org/packages) function available on the Nixpkgs
+website.
+
+(TODO: this is false without --ignore-environment, which we haven't put into the
```

Comment:
Not sure what to do here; `--ignore-environment` seems to not reset everything, and even if it did, it can't be used (directly) from direnv. I've asked again on #nix on Slack.

Reply:
The only idea I have at this point is using `flake-compat` and getting a shell.nix and doing `use nix --pure`, which _is_ supported.

Reply:
I've rewritten the section, but it feels a little bit like lying by omission if we talk a big game about how good Nix shells are for testing software interactively without unnecessary dependencies and then provide a shell setup that isn't.

The hermeticity/correctness of the rules_nixpkgs-based Bazel build itself doesn't depend on this, though, as far as I know, so I suppose this setup won't actually create any incorrect builds. We can leave it at this for a first draft until we find a better solution.